### PR TITLE
Set ReadResult and internal fields as readonly.

### LINKED ISSUE #29683: System.IO.Pipelines.ReadResult should be readonly
So it can be efficiently passed with `in`
```csharp
public readonly struct ReadResult
{
    internal readonly ReadOnlySequence<byte> _resultBuffer;
    internal readonly ResultFlags _resultFlags;
}
```

/cc @davidfowl @pakrym

### PR DESCRIPTION
**Pull request by CodeTherapist**
Fix missing period in xml summary.

closes #29683

### CODE REVIEW

**Comment by pakrym** [comment]
As discussed this needs to be perf tested before getting merged. Code was written the way it was for performance reasons.

**Comment by CodeTherapist** [comment]
@pakrym I will deliver the perf tests on the next Tuesday.

**Comment by CodeTherapist** [comment]
@benaadams I added it - how did you got that result (isn't `Build.cmd -Release` enough)?

**Comment by benaadams** [comment]
Its the packaging step; checked the error on the CI details :)

It think you need to run all the tests also to trigger it locally

**Comment by benaadams** [comment]
@dotnet-bot test please

**Comment by CodeTherapist** [comment]
@benaadams Thanks for your answer - that helps me to getting used to it and understand how it works.

**Comment by benaadams** [comment]
@dotnet-bot test

**Comment by benaadams** [comment]
@CodeTherapist CI seems to only have run 2 jobs; try closing and reopening, that might kick it back into life

**Comment by benaadams** [comment]
Seemed to work 😄

**Comment by CodeTherapist** [comment]
Yes, definitely :)

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -638,17 +637,17 @@ internal bool TryRead(out ReadResult result)
...
-                    GetReadResult(ref result);
+                    GetReadResult(out result);
                     return true;
                 }
 
                 if (_readerAwaitable.HasContinuation)
                 {
                     ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
                 }
+                result = new ReadResult();
```
Comment (pakrym) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: `= default`

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 2]
```diff
@@ -781,34 +780,29 @@ internal ReadResult GetReadAsyncResult()
...
             if (head != null)
             {
                 // Reading commit head shared with writer
-                result._resultBuffer = new ReadOnlySequence<byte>(head, _readHeadIndex, _commitHead, _commitHeadIndex - _commitHead.Start);
+                var readOnlySequence = new ReadOnlySequence<byte>(head, _readHeadIndex, _commitHead, _commitHeadIndex - _commitHead.Start);
+                result = new ReadResult(readOnlySequence, isCanceled, isCompleted);
+            }
+            else
+            {
+                result = new ReadResult(new ReadOnlySequence<byte>(), isCanceled, isCompleted);
```
Comment (pakrym) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: default

Comment (CodeTherapist) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Isn't there any way that `isCanceled` or `isCompleted` is `true`?

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Think he means `new ReadOnlySequence<byte>()` => `default`

So
```csharp
result = new ReadResult(default, isCanceled, isCompleted);
```

Comment (CodeTherapist) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Ah, Sorry, that wasn't clear -  that's fine.

File: src/System.IO.Pipelines/tests/Performance/Configurations.props [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <BuildConfigurations>
+      netstandard;
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/Performance/Configurations.props]: @ahsonkhan does benchmarking project look good to you?

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <IncludePerformanceTests>true</IncludePerformanceTests>
+    <ProjectGuid>{66AE57BA-B56C-4A1E-ACA6-7C18431D416B}</ProjectGuid>
+    <Configurations>netstandard-Debug;netstandard-Release</Configurations>
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="$(CommonTestPath)\System\Diagnostics\RemoteExecutorConsoleApp\RemoteExecutorConsoleApp.csproj">
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: I am not sure if you need this at all. I would remove it.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      <Project>{69e46a6f-9966-45a5-8945-2559fe337827}</Project>
+      <Name>RemoteExecutorConsoleApp</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\System.IO.Pipelines.Tests.csproj">
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: I am assuming this is needed for TestMemoryPool? Maybe include that file explicitly and not the whole test project?

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    </ProjectReference>
+
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: nit: remove space.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="$(CommonTestPath)\System\Diagnostics\RemoteExecutorConsoleApp\RemoteExecutorConsoleApp.csproj">
+      <Project>{69e46a6f-9966-45a5-8945-2559fe337827}</Project>
+      <Name>RemoteExecutorConsoleApp</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\System.IO.Pipelines.Tests.csproj">
+    </ProjectReference>
+        
+    <Compile Include="Perf.Pipe.cs" />
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: nit: move source file includes into a separate ItemGroup

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 8]
```diff
@@ -781,34 +780,29 @@ internal ReadResult GetReadAsyncResult()
...
-                result._resultFlags |= ResultFlags.Completed;
-            }
-
-            bool isCanceled = _readerAwaitable.ObserveCancelation();
-            if (isCanceled)
-            {
-                result._resultFlags |= ResultFlags.Canceled;
-            }
-
+            var isCompleted = _writerCompletion.IsCompletedOrThrow();
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: nit: don't use var here and below.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj (hunk 5 above)
Comment (CodeTherapist) [reply src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: Yes, `TestMemoryPool` is in `System.IO.Pipelines.Tests`. But it looks like that I could use the default `MemoryPool`. I will remove the dependency.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj (hunk 4 above)
Comment (CodeTherapist) [reply src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: It's not needed. I took the another Performance project as a template - I thought that this is a common setup.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj (hunk 6 above)
Comment (CodeTherapist) [reply src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: Done.

File: src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj (hunk 7 above)
Comment (CodeTherapist) [reply src/System.IO.Pipelines/tests/Performance/System.IO.Pipelines.Performance.Tests.csproj]: Done.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 8 above)
Comment (CodeTherapist) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Done.

**Comment by pakrym** [comment]
@CodeTherapist, thank you for your contribution.

**Pull request merged**